executor/foreign_key_test.go

```diff
@@ -681,3 +695,49 @@ func TestForeignKeyOnUpdateParentTableCheck(t *testing.T) {
 	tk.MustQuery("select id, a, b from t1 order by id").Check(testkit.Rows("1 11 21", "4 14 24", "102 12 22", "103 13 23"))
 	tk.MustQuery("select id, a, b, name from t2 order by id").Check(testkit.Rows("11 1 21 a"))
 }
+
+func TestForeignKeyOnDeleteParentTableCheck(t *testing.T) {
+	store := testkit.CreateMockStore(t)
+	tk := testkit.NewTestKit(t, store)
+	tk.MustExec("set @@global.tidb_enable_foreign_key=1")
+	tk.MustExec("set @@foreign_key_checks=1")
+	tk.MustExec("use test")
+
+	for _, ca := range foreignKeyTestCase1 {
+		tk.MustExec("drop table if exists t2;")
+		tk.MustExec("drop table if exists t1;")
+		for _, sql := range ca.prepareSQLs {
+			tk.MustExec(sql)
+		}
+		if !ca.notNull {
+			tk.MustExec("insert into t1 (id, a, b) values (1, 1, 1), (2, 2, 2), (3, 3, 3), (4, 4, 4), (5, 5, null), (6, null, 6), (7, null, null);")
+			tk.MustExec("insert into t2 (id, a, b) values (1, 1, 1), (5, 5, null), (6, null, 6), (7, null, null);;")
+
+			tk.MustExec("delete from t1 where id = 2")
+			tk.MustExec("delete from t1 where a = 3 or b = 4")
+			tk.MustExec("delete from t1 where a = 5 or b = 6 or a is null or b is null;")
+			tk.MustGetDBError("delete from t1 where id = 1", plannercore.ErrRowIsReferenced2)
+			tk.MustQuery("select id, a, b from t1 order by id").Check(testkit.Rows("1 1 1"))
```

Comment:
Should this test the case that within the current transaction, rows are being modified which will then satisfy the constraint. i.e.

```sql
mysql> delete from t1 where id = 1;
ERROR 1451 (23000): Cannot delete or update a parent row: a foreign key constraint fails (`test`.`t2`, CONSTRAINT `fk` FOREIGN KEY (`a`, `b`) REFERENCES `t1` (`a`, `b`))
mysql> delete from t2 where a=1;
Query OK, 1 row affected (0.01 sec)

mysql> delete from t1 where id = 1;
Query OK, 1 row affected (0.01 sec)
```

(This is against your branch - it looks to be working correctly.)

Reply:
I found a fun test-case where TiDB differs from MySQL (using innodb with default RR):

```sql
# setup
drop table if exists t1, t2;
create table t1 (id int, a int, b int,  unique index(id), unique index(a, b));
create table t2 (b int, name varchar(10), a int, id int, unique index(id), unique index (a,b), foreign key fk(a, b) references t1(a, b));
insert into t1 (id, a, b) values (1, 1, 1), (2, 2, 2), (3, 3, 3), (4, 4, 4), (5, 5, null), (6, null, 6), (7, null, null);
insert into t2 (id, a, b) values (1, 1, 1), (5, 5, null), (6, null, 6), (7, null, null);

session1> start transaction;
session1> delete from t1 where id = 9999; # ensure snapshot is started

session2> delete from t2 where a = 1; # delete in auto commit
session1> delete from t1 where id = 1; # success in MySQL; fails in TiDB
```

I believe this is because for modifications MySQL is effectively RC not RR. TiDB is using SI still, so the delete fails even though the row no longer exists.

Reply:
I submitted an issue about this problem here: https://github.com/pingcap/tidb/issues/38354

Look like the foreign key constraint check can't work well in optimistic transactions, but work in pessimistic transaction, I already add a test case here: https://github.com/pingcap/tidb/pull/38222/files#diff-a6a940d5e03235f9d1e12c25e69801c36db3154ead88bba7d62a1282bf3407e0R428

Reply:
The problem of foreign key constraint check doesn't work well won't be fixed in this PR, I will fix it in another PR.